allennlp/modules/conditional_random_field.py

```diff
@@ -9,86 +9,135 @@
+            from_tag = from_label
+            from_entity = ""
+        else:
+            from_tag = from_label[0]
+            from_entity = from_label[1:]
+        for to_label_index, to_label in labels_with_boundaries:
+            if to_label in ("START", "END"):
+                to_tag = to_label
+                to_entity = ""
+            else:
+                to_tag = to_label[0]
+                to_entity = to_label[1:]
+            if is_transition_allowed(constraint_type, from_tag, from_entity,
+                                     to_tag, to_entity):
+                allowed.append((from_label_index, to_label_index))
+    return allowed
 
-                is_allowed = any([
-                        # Can always transition to O or B-x
-                        to_bio in ('O', 'B'),
-                        # Can only transition to I-x from B-x or I-x
-                        to_bio == 'I' and from_bio in ('B', 'I') and from_entity == to_entity
-                ])
 
-                if is_allowed:
-                    allowed.append((i, j))
+def is_transition_allowed(constraint_type: str,
+                          from_tag: str,
+                          from_entity: str,
+                          to_tag: str,
+                          to_entity: str):
+    """
+    Given a constraint type and strings ``from_tag`` and ``to_tag`` that
+    represent the origin and destination of the transition, return whether
+    the transition is allowed under the given constraint type.
 
-        # start transitions
-        for i, (to_bio, *to_entity) in tokens.items():
-            if to_bio in ('O', 'B'):
-                allowed.append((start_tag, i))
+    Parameters
+    ----------
+    constraint_type : ``str``, required
+        Indicates which constraint to apply. Current choices are
+        "BIO", "IOB1", and BIOUL".
+    from_tag : ``str``, required
+        The tag that the transition originates from. For example, if the
+        label is ``I-PER``, the ``from_tag`` is ``I``.
+    from_entity: ``str``, required
+        The entity corresponding to the ``from_tag``. For example, if the
+        label is ``I-PER``, the ``from_entity`` is ``PER``.
+    to_tag : ``str``, required
+        The tag that the transition leads to. For example, if the
+        label is ``I-PER``, the ``to_tag`` is ``I``.
+    to_entity: ``str``, required
+        The entity corresponding to the ``to_tag``. For example, if the
+        label is ``I-PER``, the ``to_entity`` is ``PER``.
 
-        # end transitions
-        for i, (from_bio, *from_entity) in tokens.items():
-            if from_bio in ('O', 'B', 'I'):
-                allowed.append((i, end_tag))
+    Returns
+    -------
+    ``bool``
+        Whether the transition is allowed under the given ``constraint_type``.
+    """
+    if to_tag == "START" or from_tag == "END":
+        # Cannot transition into START or from END
+        return False
+    if from_tag == "START":
```

Comment:
Nit: you've grouped "START" and "END" together here, and I now have three places to look in the code to see all of the "BIOUL" constraints.  Seems more coherent to instead have a `if from_tag == "START"` block inside of each `if constraint_type == ...` block below.